velox/exec/Unnest.cpp

```diff
@@ -284,32 +335,77 @@ const Unnest::UnnestChannelEncoding Unnest::generateEncodingForChannel(
 }
 
 VectorPtr Unnest::generateOrdinalityVector(const RowRange& range) {
+  VELOX_DCHECK_GT(range.numInputRows, 0);
+
   auto ordinalityVector = BaseVector::create<FlatVector<int64_t>>(
       BIGINT(), range.numInnerRows, pool());
-
   // Set the ordinality at each result row to be the index of the element in
   // the original array (or map) plus one.
   auto* rawOrdinality = ordinalityVector->mutableRawValues();
+  const bool hasEmptyUnnestValue =
+      withEmptyUnnestValue_ && range.hasEmptyUnnestValue;
+  if (!hasEmptyUnnestValue) {
+    range.forEachRow(
+        [&](vector_size_t /*row*/, vector_size_t start, vector_size_t size) {
+          std::iota(rawOrdinality, rawOrdinality + size, start + 1);
+          rawOrdinality += size;
+        },
+        rawMaxSizes_,
+        firstInnerRowStart_);
+  } else {
+    range.forEachRow(
+        [&](vector_size_t /*row*/, vector_size_t start, vector_size_t size) {
+          if (FOLLY_LIKELY(size > 0)) {
+            std::iota(rawOrdinality, rawOrdinality + size, start + 1);
+            rawOrdinality += size;
+          } else {
+            // Set ordinality to 0 for output row with empty unnest value.
+            VELOX_DCHECK_EQ(size, 0);
+            *rawOrdinality++ = 0;
+          }
+        },
+        rawMaxSizes_,
+        firstInnerRowStart_);
+  }
+  return ordinalityVector;
+}
 
+VectorPtr Unnest::generateEmptyUnnestValueVector(const RowRange& range) {
+  VELOX_CHECK(withEmptyUnnestValue_);
   VELOX_DCHECK_GT(range.numInputRows, 0);
 
+  if (!range.hasEmptyUnnestValue) {
+    return BaseVector::createConstant(
+        BOOLEAN(), false, range.numInnerRows, pool());
+  }
+
+  auto emptyVector = BaseVector::create<FlatVector<bool>>(
+      BOOLEAN(), range.numInnerRows, pool());
+  // Set each output row has empty unnest values or not.
+  auto* const rawEmpty = emptyVector->mutableRawValues<uint64_t>();
```

Comment:
It might be cleaner to assume that most rows will be false and allocate values buffer that's 'false' in all rows. Then, flip the value for rows that are empty.